cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -0,0 +1,431 @@
+import co.cask.cdap.report.proto.ReportContent;
+import co.cask.cdap.report.proto.ReportGenerationInfo;
+import co.cask.cdap.report.proto.ReportGenerationRequest;
+import co.cask.cdap.report.proto.ReportList;
+import co.cask.cdap.report.proto.ReportStatus;
+import co.cask.cdap.report.proto.ReportStatusInfo;
+import co.cask.cdap.report.proto.ValueFilter;
+import co.cask.cdap.report.util.ReportField;
+import co.cask.cdap.report.util.ReportIds;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.ByteStreams;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonSyntaxException;
+import com.google.gson.reflect.TypeToken;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.SparkSession;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+
+/**
+ * A Spark program for generating reports, querying for report statuses, and reading reports.
+ */
+public class ReportGenerationSpark extends AbstractExtendedSpark implements JavaSparkMain {
+  private static final Logger LOG = LoggerFactory.getLogger(ReportGenerationSpark.class);
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Filter.class, new FilterDeserializer())
+    .create();
+
+  @Override
+  protected void configure() {
+    setMainClass(ReportGenerationSpark.class);
+    addHandlers(new ReportSparkHandler());
+  }
+
+  @Override
+  public void run(JavaSparkExecutionContext sec) throws Exception {
+    JavaSparkContext jsc = new JavaSparkContext();
+  }
+
+  /**
+   * A {@link SparkHttpServiceHandler} generating reports, querying for report statuses, and reading reports.
+   */
+  public static final class ReportSparkHandler extends AbstractSparkHttpServiceHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ReportSparkHandler.class);
+    private static final Type REPORT_GENERATION_REQUEST_TYPE = new TypeToken<ReportGenerationRequest>() {
+    }.getType();
+    private static final String DEFAULT_LIMIT = "10000";
+    private SparkSession sparkSession;
+
+    public static final String READ_LIMIT = "readLimit";
```

Comment:
Why all these are `public` instead of `private? I don't see they are used outside of this class.